src/nanoemoji/bitmap_tables.py

```diff
@@ -0,0 +1,265 @@
+def _ppem(config: FontConfig, advance: int) -> int:
+    return int(_width_in_pixels(config) * config.upem / advance)
+
+
+def _advance(ttfont: ttLib.TTFont, color_glyphs: Sequence[ColorGlyph]) -> int:
+    # let's go ahead and fail miserably if advances are not all the same
+    # proportional bitmaps can wait for a second pass :)
+    advances = {
+        ttfont["hmtx"].metrics[ttfont.getGlyphName(c.glyph_id)][0] for c in color_glyphs
+    }
+    assert len(advances) == 1, "Proportional bitmaps not supported yet"
+    return next(iter(advances))
+
+
+def _image_bytes(color_glyph: ColorGlyph) -> bytes:
+    with open(color_glyph.filename, "rb") as f:
+        return f.read()
+
+
+def _cbdt_record_size(image_format: int, image_data: bytes) -> int:
+    assert image_format == _CBDT_SMALL_METRIC_PNGS, "Unrecognized format"
+    return _CBDT_SMALL_METRIC_PNG_HEADER_SIZE + len(image_data)
+
+
+def _cbdt_bitmapdata_offsets(
+    image_format: int, color_glyphs: Sequence[ColorGlyph]
+) -> List[Tuple[int, int]]:
+    # TODO is this right? - feels dumb. But ... compile crashes if locations are unknown.
+    offsets = []
+    offset = _CBDT_HEADER_SIZE
+    for color_glyph in color_glyphs:
+        offsets.append(offset)
+        offset += _cbdt_record_size(image_format, _image_bytes(color_glyph))
+    offsets.append(offset)  # capture end of stream
+    return list(zip(offsets, offsets[1:]))
+
+
+def _cbdt_bitmap_data(
+    config: FontConfig, metrics: BitmapMetrics, image_data: bytes
+) -> CbdtBitmapFormat17:
+
+    bitmap_data = CbdtBitmapFormat17(b"", None)
+    bitmap_data.metrics = SmallGlyphMetrics()
+    bitmap_data.metrics.height = config.bitmap_resolution
+    bitmap_data.metrics.width = config.bitmap_resolution
+    # center within advance
+    bitmap_data.metrics.BearingX = metrics.x_offset
+    bitmap_data.metrics.BearingY = metrics.y_offset
+    bitmap_data.metrics.Advance = _width_in_pixels(config)
+    bitmap_data.imageData = image_data
+    return bitmap_data
+
+
+def make_sbix_table(
+    config: FontConfig,
+    ttfont: ttLib.TTFont,
+    color_glyphs: Sequence[ColorGlyph],
+):
+
+    sbix = ttLib.newTable("sbix")
+    ttfont[sbix.tableTag] = sbix
+
+    ppem = _ppem(config, _advance(ttfont, color_glyphs))
+
+    strike = SbixStrike()
+    strike.ppem = ppem
+    strike.resolution = 72  # pixels per inch
+    sbix.strikes[strike.ppem] = strike
+
+    metrics = BitmapMetrics.create(config, strike.ppem)
+
+    for color_glyph in color_glyphs:
+        # TODO: if we've seen these bytes before set graphicType "dupe", referenceGlyphName <name of glyph>
+        image_data = _image_bytes(color_glyph)
+
+        glyph_name = ttfont.getGlyphName(color_glyph.glyph_id)
+        glyph = SbixGlyph(
+            graphicType="png",
+            glyphName=glyph_name,
+            imageData=image_data,
+            originOffsetX=metrics.x_offset,
+            originOffsetY=metrics.line_height - metrics.y_offset,
```

Comment:
I get weird vertical positioning in Safari Mac with your originOffsetY code:

<img width="277" alt="Screenshot 2022-01-25 at 11 47 09" src="https://user-images.githubusercontent.com/6939968/150972099-9e09e6d2-fa3d-4d1a-bed1-9f2f74d4428f.png">

If I flip it like so, I get the expected placement:

```suggestion
            originOffsetY=metrics.y_offset - metrics.line_height,
```

<img width="278" alt="Screenshot 2022-01-25 at 11 51 23" src="https://user-images.githubusercontent.com/6939968/150972116-5a52abbf-8c5a-4b3f-b190-905ccfe25a0c.png">

Also using line_ascent intead of y_offset looks ok (the two ought to be equal because we stretch to fit the vertical distance and only center horizontally -- actually if y_offset and line_ascent don't match exactly I'd count that as a bug):
```suggestion
            originOffsetY=metrics.line_ascent - metrics.line_height,
```

Reply:
uh-oh... I think we found a bug in sbix Chrome implementation..
Chrome on Linux (first screenshot below) doesn't match Safari on Mac's treatment of originOffsetY (second screenshot).

<img width="417" alt="Screenshot 2022-01-25 at 12 18 05" src="https://user-images.githubusercontent.com/6939968/150976022-9e71b5e0-e800-4a1c-a627-f7fc82ea8bc9.png">

<img width="277" alt="Screenshot 2022-01-25 at 11 47 09" src="https://user-images.githubusercontent.com/6939968/150976067-65f9be0a-7d31-4a04-a58f-c9d03e5f3854.png">

This is using your current code for `originOffsetY=metrics.line_height - metrics.y_offset`.
If I invert that to `metrics.y_offset - metrics.line_height`, I fix Safari Mac but break Chrome Linux!
I attached the test font and static html page here:
[Archive.zip](https://github.com/googlefonts/nanoemoji/files/7933523/Archive.zip)

~~~

Actually, we maybe found two bugs: unlike Chrome on Linux, the very same sbix font and html page renders all blank in Chrome **on Mac**!

![image](https://user-images.githubusercontent.com/6939968/150976786-7d38704b-fc2f-4c86-9d77-9ffdffbcf970.png)

Reply:
as for the second bug (sbix test font not displaying in Chrome for Mac), it's because nanoemoji is writing empty glyphs in glyf table -- ever since ClipBox were added to COLRv1 it no longer emits the two invisible glyf contour points that delimit the bounding box of the color glyph. Looks like Safari Mac doesn't care if the glyphs are empty and displays the color glyphs nonetheless, but _Chrome Mac won't display the sbix color glyphs unless the respective b/w glyf glyphs contain an outline whose bbox is enough to contain the color glyph_. So for sbix at least we need those back.

Reply:
The sbix opentype spec has for `originOffsetY` (emphasis mine)

> The vertical (y-axis) position of the **bottom** edge of the bitmap graphic in relation to the glyph design space origin.

then it follows with:

> The originOffsetX and originOffsetY values give the placement of the bitmap graphic in relation to the standard coordinate system of the glyph design space.

So, Safari Mac has it correct, and Chrome has it wrong. In this test font, the originOffsetY is -25, which means the bottom of the bitmap should be shifted 25 units _down_ (not up!) in relation to the glyph origin.

Reply:
I found this https://github.com/MicrosoftDocs/typography-issues/issues/191

Reply:
The OT sbix spec also says:

> If there is a glyph contour, the glyph design space origin for the graphic is placed at the lower left corner of the glyph bounding box (xMin, yMin)

So, since we need to change nanoemoji to emit the bbox lower-left and top-right points anyway (otherwise Chrome Mac displays all blank) and hence we have to have some non-empty glyf contour that fully covers the bitmap width/height, then we could just set originOffsetX/Y == 0,  because we want the bitmap to be placed exactly "at the lower left corner of the glyph bounding box (xMin, yMin)". Hopefully everything aligns to full integers, unless upem (or ascender - descender, actually, we use that to scale the bitmap) and bitmap pixel size are not exactly divisible.

Reply:
FreeType still hasn't been fixed, it's applying originOffsetY in the wrong direction (y-down), and it's not taking into account whether a glyf entry esists for the color glyph (in which case the glyf bbox lower left point should be used as the origin for the originOffsetX/Y):
https://gitlab.freedesktop.org/freetype/freetype/-/issues/998

This is the line where it sets the `FT_Glyph_Metrics.horiBearingY` (distance from baseline to the top of the bitmap) from sbix `originOffsetY` (note the minus sign): https://gitlab.freedesktop.org/freetype/freetype/-/blob/master/src/sfnt/ttsbit.c#L1590

`metrics->horiBearingY = (FT_Short)( -originOffsetY + metrics->height );`

Now I wonder, what if we do add the non-empty glyf entry with lower-left and top-right invisible points (to make the bitmap draw on Mac) and set originOffsetX/Y to 0 accordingly.. then FreeType would still appear incorrect because it ignores the glyf entry and just uses the baseline y=0 as the origin for originOffsetY instead of the bbox.yMin... 🤔

Reply:
Anyway, if nanoemoji ships with any sbix support, the fonts should at least work on the Apple platform which is the canonical implementation. It may be impossible to make them work reliably everywhere because of the legacy issues above.

Reply:
according to Ben, empty glyf entries may also still be an issue on Safari Mac: https://github.com/harfbuzz/harfbuzz/issues/2679#issuecomment-1021479899

Reply:
Per https://github.com/harfbuzz/harfbuzz/issues/2679#issuecomment-1021479899 it appears we need a glyf entry even for Safari